### PR TITLE
Implement `cudf::cast` for `decimal32/64` to/from integer and floating point

### PR DESCRIPTION
**Pull request by codereport**
This PR resolves https://github.com/rapidsai/cudf/issues/6486.
This PR resolves a part of https://github.com/rapidsai/cudf/issues/3556.

A thought: not really sure why `cudf::cast` is in unary ops. It is a binary op (aka imo makes more sense for it to be in its own separate file similar to `cudf::round`).

### To do list:
* [x] Integer to `fixed_point`
    * [x] Add basic unit test
    * [x] Add implementation
    * [x] Add comprehensive unit tests
* [x] Floating point to `fixed_point`
    * [x] Add basic unit test
    * [x] Add implementation
    * [x] Add comprehensive unit tests
* [x] `fixed_point` to floating point
    * [x] Add basic unit test
    * [x] Add implementation
    * [x] Add comprehensive unit tests
* [x] `fixed_point` to integer
    * [x] Add basic unit test
    * [x] Add implementation
    * [x] Add comprehensive unit tests
* [x] Update documentation
* ~~Move `cudf::cast` tests to a new test file ?~~ will do in a follow up PR

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by harrism** [comment]
@codereport why do you say casting is a binary op? It operates on a single input, and produces a single output.

**Comment by codereport** [comment]
> @codereport why do you say casting is a binary op? It operates on a single input, and produces a single output. And from what I've seen cast_ops.cu is its own file.

> It operates on a single input

That isn't true, it takes two inputs.
```cpp
std::unique_ptr<column> cast(column_view const& input, data_type out_type);
```
`data_type out_type` is the second input.

**Comment by codereport** [comment]
@harrism Furthermore, if you compare the APIS in `unary.hpp` you will see that `cudf::cast` is an outlier (ignoring the `rmm::mr::device_memory_resource* mr` parameter):

```cpp
std::unique_ptr<cudf::column> is_null   (cudf::column_view const& input);
std::unique_ptr<cudf::column> is_valid  (cudf::column_view const& input);
std::unique_ptr<column>       is_nan    (cudf::column_view const& input);
std::unique_ptr<column>       is_not_nan(cudf::column_view const& input);
std::unique_ptr<column>       cast      (cudf::column_view const& input, data_type out_type);
```

**Comment by harrism** [comment]
OK, but binary ops in cudf take two arguments of the same type. E.g two columns. Column and scalar of same type.

We don't  call every cuDF function that takes a column plus a single configuration argument a binary operation.

And if you rewrote it as cast_to_int, cast_to_float they would be unary ops, just like add_two, (which takes a number and adds 2) would be a unary op.

**Comment by codereport** [comment]
> And if you rewrote it as cast_to_int, cast_to_float they would be unary ops, just like add_two, (which takes a number and adds 2) would be a unary op.

That is a good point. I guest I don't really consider `static_cast<int32_t>(...)` a binary op.

**Comment by harrism** [comment]
Also, I don't necessarily disagree that casting is *not* a unary op, I just don't think it's necessarily a binary op. I think the key is that it is an operation that might be inserted as a node into an abstract syntax tree. So we ultimately need it to work in any AST we support. And I do agree that maybe it makes sense to separate the `cast` API into its own header.

File: cpp/src/unary/cast_ops.cu [hunk 1]
```diff
@@ -128,19 +177,72 @@ struct dispatch_unary_cast_to {
...
+            typename std::enable_if_t<cudf::is_fixed_point<SourceT>() &&
+                                      cudf::is_numeric<TargetT>()>* = nullptr>
+  std::unique_ptr<column> operator()(data_type type,
+                                     rmm::mr::device_memory_resource* mr,
+                                     cudaStream_t stream)
+  {
+    auto const size = input.size();
+    auto output     = std::make_unique<column>(type,
+                                           size,
+                                           rmm::device_buffer{size * cudf::size_of(type), 0, mr},
```
Comment (davidwendt) [inline cpp/src/unary/cast_ops.cu:190]:
```suggestion
                                           rmm::device_buffer{size * cudf::size_of(type), stream, mr},
```

File: cpp/src/unary/cast_ops.cu [hunk 2]
```diff
@@ -128,19 +177,72 @@ struct dispatch_unary_cast_to {
...
+                                           copy_bitmask(input, 0, mr),
```
Comment (davidwendt) [inline cpp/src/unary/cast_ops.cu:191]:
```suggestion
                                           copy_bitmask(input, stream, mr),
```

File: cpp/src/unary/cast_ops.cu [hunk 3]
```diff
@@ -94,24 +95,72 @@ struct unary_cast {
...
                                      rmm::mr::device_memory_resource* mr,
                                      cudaStream_t stream)
   {
-    auto size   = input.size();
-    auto output = std::make_unique<column>(type,
+    auto const size = input.size();
+    auto output     = std::make_unique<column>(type,
                                            size,
                                            rmm::device_buffer{size * cudf::size_of(type), 0, mr},
                                            copy_bitmask(input, 0, mr),
```
Comment (davidwendt) [inline cpp/src/unary/cast_ops.cu:166]: Should be passing the `stream` parameter in these two lines instead of 0.

File: cpp/src/unary/cast_ops.cu [hunk 4]
```diff
@@ -128,19 +177,72 @@ struct dispatch_unary_cast_to {
...
+            typename std::enable_if_t<cudf::is_numeric<SourceT>() &&
+                                      cudf::is_fixed_point<TargetT>()>* = nullptr>
+  std::unique_ptr<column> operator()(data_type type,
+                                     rmm::mr::device_memory_resource* mr,
+                                     cudaStream_t stream)
+  {
+    auto const size = input.size();
+    auto output     = std::make_unique<column>(type,
+                                           size,
+                                           rmm::device_buffer{size * cudf::size_of(type), 0, mr},
```
Comment (davidwendt) [inline cpp/src/unary/cast_ops.cu:218]:
```suggestion
                                           rmm::device_buffer{size * cudf::size_of(type), stream, mr},
```

File: cpp/src/unary/cast_ops.cu [hunk 5]
```diff
@@ -128,19 +177,72 @@ struct dispatch_unary_cast_to {
...
+                                           copy_bitmask(input, 0, mr),
```
Comment (davidwendt) [inline cpp/src/unary/cast_ops.cu:219]:
```suggestion
                                           copy_bitmask(input, stream, mr),
```

**Review by davidwendt — CHANGES_REQUESTED** [review]
Just missing a few of stream parameters.

File: cpp/src/unary/cast_ops.cu (hunk 1 above)
Comment (codereport) [reply cpp/src/unary/cast_ops.cu:190]: Done.

File: cpp/src/unary/cast_ops.cu (hunk 2 above)
Comment (codereport) [reply cpp/src/unary/cast_ops.cu:191]: Done.

File: cpp/src/unary/cast_ops.cu (hunk 3 above)
Comment (codereport) [reply cpp/src/unary/cast_ops.cu:166]: Done.

File: cpp/src/unary/cast_ops.cu (hunk 4 above)
Comment (codereport) [reply cpp/src/unary/cast_ops.cu:218]: Done.

File: cpp/src/unary/cast_ops.cu (hunk 5 above)
Comment (codereport) [reply cpp/src/unary/cast_ops.cu:219]: Done.

**Review by davidwendt — APPROVED** [review]

**Review by trxcllnt — APPROVED** [review]

**Pull request merged**